diff --git a/plugins/a11y-keyboard/gsd-a11y-keyboard-manager.c b/plugins/a11y-keyboard/gsd-a11y-keyboard-manager.c
index 07bb251..5e338c5 100644
--- a/plugins/a11y-keyboard/gsd-a11y-keyboard-manager.c
+++ b/plugins/a11y-keyboard/gsd-a11y-keyboard-manager.c
@@ -1023,6 +1023,11 @@ gsd_a11y_keyboard_manager_stop (GsdA11yKeyboardManager *manager)
 
         g_debug ("Stopping a11y_keyboard manager");
 
+        if (p->start_idle_id != 0) {
+                g_source_remove (p->start_idle_id);
+                p->start_idle_id = 0;
+        }
+
         if (p->device_manager != NULL) {
                 g_signal_handler_disconnect (p->device_manager, p->device_added_id);
                 p->device_manager = NULL;
@@ -1034,6 +1039,7 @@ gsd_a11y_keyboard_manager_stop (GsdA11yKeyboardManager *manager)
         }
 
         if (p->settings != NULL) {
+                g_signal_handlers_disconnect_by_func (p->settings, keyboard_callback, manager);
                 g_object_unref (p->settings);
                 p->settings = NULL;
         }
