Workaround until GNOME#631989 is fixed.

Index: gnome-settings-daemon-3.0.2/data/org.gnome.settings-daemon.peripherals.gschema.xml.in.in
===================================================================
--- gnome-settings-daemon-3.0.2.orig/data/org.gnome.settings-daemon.peripherals.gschema.xml.in.in	2011-06-04 17:41:23.778543311 +0200
+++ gnome-settings-daemon-3.0.2/data/org.gnome.settings-daemon.peripherals.gschema.xml.in.in	2011-06-04 17:44:16.743388902 +0200
@@ -95,6 +95,11 @@
       <summary>Remember NumLock state</summary>
       <description>When set to true, GNOME will remember the state of the NumLock LED between sessions.</description>
     </key>
+    <key name="numlock-state" type="b">
+      <default>false</default>
+      <summary>State of the NumLock LED</summary>
+      <description>This value holds the last remembered state for the NumLock LED.</description>
+    </key>
   </schema>
   <schema gettext-domain="@GETTEXT_PACKAGE@" id="org.gnome.settings-daemon.peripherals.mouse" path="/org/gnome/settings-daemon/peripherals/mouse/">
     <key name="locate-pointer" type="b">
Index: gnome-settings-daemon-3.0.2/plugins/keyboard/gsd-keyboard-manager.c
===================================================================
--- gnome-settings-daemon-3.0.2.orig/plugins/keyboard/gsd-keyboard-manager.c	2011-06-04 17:37:48.789492285 +0200
+++ gnome-settings-daemon-3.0.2/plugins/keyboard/gsd-keyboard-manager.c	2011-06-04 17:46:00.287895098 +0200
@@ -119,7 +119,7 @@ gsd_keyboard_get_hostname_key (void)
                 return NULL;
         }
 #endif
-        return NULL;
+        return g_strdup("numlock-state");
 }
 
 #ifdef HAVE_X11_EXTENSIONS_XKB_H
