diff -Nur -x '*.orig' -x '*~' gnome-settings-daemon-2.27.4/plugins/xrdb/gsd-xrdb-manager.c gnome-settings-daemon-2.27.4.new/plugins/xrdb/gsd-xrdb-manager.c
--- gnome-settings-daemon-2.27.4/plugins/xrdb/gsd-xrdb-manager.c	2009-05-03 19:15:30.000000000 +0100
+++ gnome-settings-daemon-2.27.4.new/plugins/xrdb/gsd-xrdb-manager.c	2009-07-19 23:03:26.381825664 +0100
@@ -42,7 +42,7 @@
 
 #define GSD_XRDB_MANAGER_GET_PRIVATE(o) (G_TYPE_INSTANCE_GET_PRIVATE ((o), GSD_TYPE_XRDB_MANAGER, GsdXrdbManagerPrivate))
 
-#define SYSTEM_AD_DIR    DATADIR "/xrdb"
+#define SYSTEM_AD_DIR    "/etc/gnome/config"
 #define GENERAL_AD       SYSTEM_AD_DIR "/General.ad"
 #define USER_AD_DIR      ".gnome2/xrdb"
 #define USER_X_RESOURCES ".Xresources"
