Change the directory for xrdb data.

Index: gnome-settings-daemon-2.26.0/plugins/xrdb/gsd-xrdb-manager.c
===================================================================
--- gnome-settings-daemon-2.26.0.orig/plugins/xrdb/gsd-xrdb-manager.c	2009-01-05 17:26:26.000000000 +0100
+++ gnome-settings-daemon-2.26.0/plugins/xrdb/gsd-xrdb-manager.c	2009-04-11 13:36:54.898047572 +0200
@@ -42,7 +42,7 @@
 
 #define GSD_XRDB_MANAGER_GET_PRIVATE(o) (G_TYPE_INSTANCE_GET_PRIVATE ((o), GSD_TYPE_XRDB_MANAGER, GsdXrdbManagerPrivate))
 
-#define SYSTEM_AD_DIR    DATADIR "/xrdb"
+#define SYSTEM_AD_DIR    "/etc/gnome/config"
 #define GENERAL_AD       SYSTEM_AD_DIR "/General.ad"
 #define USER_AD_DIR      ".gnome2/xrdb"
 #define USER_X_RESOURCES ".Xresources"
