Index: gnome-settings-daemon-2.29.5/configure.ac
===================================================================
--- gnome-settings-daemon-2.29.5.orig/configure.ac	2010-01-12 15:52:55.000000000 +0000
+++ gnome-settings-daemon-2.29.5/configure.ac	2010-01-17 19:16:55.000000000 +0000
@@ -68,6 +68,8 @@
         gnome-desktop-2.0 >= $GNOME_DESKTOP_REQUIRED_VERSION
         gio-2.0 >= $GIO_REQUIRED_VERSION
         dbus-glib-1 >= $DBUS_GLIB_REQUIRED_VERSION
+        xext
+        x11
 )
 
 GSD_PLUGIN_LDFLAGS="-export_dynamic -module -avoid-version -no-undefined"
