Index: gnome-settings-daemon-2.24.1/configure.ac
===================================================================
--- gnome-settings-daemon-2.24.1.orig/configure.ac	2008-12-27 12:23:20.551491461 +0100
+++ gnome-settings-daemon-2.24.1/configure.ac	2008-12-27 12:27:09.064807679 +0100
@@ -68,6 +68,8 @@ PKG_CHECK_MODULES(SETTINGS_PLUGIN,
         gio-2.0 >= $GIO_REQUIRED_VERSION
         libglade-2.0 >= 2.0.0
         dbus-glib-1 >= $DBUS_GLIB_REQUIRED_VERSION
+        xext
+        x11
 )
 
 PKG_CHECK_MODULES(GNOME, libgnome-2.0)
