Index: gnome-settings-daemon-2.30.1/configure.ac
===================================================================
--- gnome-settings-daemon-2.30.1.orig/configure.ac	2010-04-26 14:19:03.000000000 +0200
+++ gnome-settings-daemon-2.30.1/configure.ac	2010-04-27 19:30:51.998031911 +0200
@@ -69,6 +69,8 @@ PKG_CHECK_MODULES(SETTINGS_PLUGIN,
         gnome-desktop-2.0 >= $GNOME_DESKTOP_REQUIRED_VERSION
         gio-2.0 >= $GIO_REQUIRED_VERSION
         dbus-glib-1 >= $DBUS_GLIB_REQUIRED_VERSION
+        xext
+        x11
 )
 
 GSD_PLUGIN_LDFLAGS="-export_dynamic -module -avoid-version -no-undefined"
