Index: gnome-settings-daemon-2.26.1/configure.ac
===================================================================
--- gnome-settings-daemon-2.26.1.orig/configure.ac	2009-05-29 16:19:04.081105324 +0200
+++ gnome-settings-daemon-2.26.1/configure.ac	2009-05-29 16:19:32.437481894 +0200
@@ -11,6 +11,7 @@ AC_INIT([gnome-settings-daemon],
 AC_CONFIG_SRCDIR([gnome-settings-daemon/gnome-settings-manager.c])
 
 AM_INIT_AUTOMAKE([1.9 no-dist-gzip dist-bzip2 tar-ustar])
+AM_MAINTAINER_MODE
 
 AC_STDC_HEADERS
 AC_PROG_CXX
