diff --git a/data/org.gnome.settings-daemon.plugins.media-keys.gschema.xml.in.in b/data/org.gnome.settings-daemon.plugins.media-keys.gschema.xml.in.in
index 85438c5..c3ad6c0 100644
--- a/data/org.gnome.settings-daemon.plugins.media-keys.gschema.xml.in.in
+++ b/data/org.gnome.settings-daemon.plugins.media-keys.gschema.xml.in.in
@@ -80,6 +80,11 @@
       <_summary>Stop playback</_summary>
       <_description>Binding to stop playback.</_description>
     </key>
+    <key name="video-out" type="s">
+      <default>'&lt;Super&gt;p'</default>
+      <_summary>Video output</_summary>
+      <_description>Binding to switch the video output device.</_description>
+    </key>
     <key name="volume-down" type="s">
       <default>'XF86AudioLowerVolume'</default>
       <_summary>Volume down</_summary>
diff --git a/plugins/media-keys/acme.h b/plugins/media-keys/acme.h
index 54ec19b..028609a 100644
--- a/plugins/media-keys/acme.h
+++ b/plugins/media-keys/acme.h
@@ -51,11 +51,11 @@ typedef enum {
         STOP_KEY,
         PREVIOUS_KEY,
         NEXT_KEY,
+        VIDEO_OUT_KEY,
         REWIND_KEY,
         FORWARD_KEY,
         REPEAT_KEY,
         RANDOM_KEY,
-        VIDEO_OUT_KEY,
         VIDEO_OUT2_KEY,
         ROTATE_VIDEO_KEY,
         MAGNIFIER_KEY,
@@ -110,12 +110,12 @@ static struct {
         { STOP_KEY, "stop", NULL, NULL },
         { PREVIOUS_KEY, "previous", NULL, NULL },
         { NEXT_KEY, "next", NULL, NULL },
+        { VIDEO_OUT_KEY, "video-out", NULL, NULL },
         /* Those are not configurable in the UI */
         { REWIND_KEY, NULL, "XF86AudioRewind", NULL },
         { FORWARD_KEY, NULL, "XF86AudioForward", NULL },
         { REPEAT_KEY, NULL, "XF86AudioRepeat", NULL },
         { RANDOM_KEY, NULL, "XF86AudioRandomPlay", NULL},
-        { VIDEO_OUT_KEY, NULL, "<Super>p", NULL },
         /* Key code of the XF86Display key (Fn-F7 on Thinkpads, Fn-F4 on HP machines, etc.) */
         { VIDEO_OUT2_KEY, NULL, "XF86Display", NULL },
         /* Key code of the XF86RotateWindows key (present on some tablets) */
