diff --git a/plugins/cursor/gsd-cursor-manager.c b/plugins/cursor/gsd-cursor-manager.c
index 611392b..563169c 100644
--- a/plugins/cursor/gsd-cursor-manager.c
+++ b/plugins/cursor/gsd-cursor-manager.c
@@ -260,6 +260,9 @@ gsd_cursor_manager_start (GsdCursorManager *manager,
         g_debug ("Starting cursor manager");
         gnome_settings_profile_start (NULL);
 
+        g_debug ("Not starting curor manager, assuming old X server");
+        return TRUE;
+
         if (supports_cursor_xfixes () == FALSE) {
                 g_debug ("XFixes cursor extension not available, will not hide the cursor");
                 return FALSE;
