Index: gnome-settings-daemon-2.29.90/gnome-settings-daemon/Makefile.am
===================================================================
--- gnome-settings-daemon-2.29.90.orig/gnome-settings-daemon/Makefile.am	2010-02-16 13:12:25.420341287 +0100
+++ gnome-settings-daemon-2.29.90/gnome-settings-daemon/Makefile.am	2010-02-16 13:12:34.508340382 +0100
@@ -37,7 +37,8 @@
 gsddir = $(libexecdir)
 
 gsd_PROGRAMS = \
-	gnome-settings-daemon
+	gnome-settings-daemon \
+	gnome-update-wallpaper-cache
 
 gnome-settings-manager-glue.h: gnome-settings-manager.xml Makefile.am
 	dbus-binding-tool --prefix=gnome_settings_manager --mode=glib-server $< > $@
@@ -80,6 +81,20 @@
 	libgsd-profile.la		\
 	$(SETTINGS_DAEMON_LIBS)
 
+gnome_update_wallpaper_cache_SOURCES =	\
+	gnome-update-wallpaper-cache.c	
+
+gnome_update_wallpaper_cache_CFLAGS =	\
+	$(SETTINGS_PLUGIN_CFLAGS)	\
+        $(AM_CFLAGS)
+
+gnome_update_wallpaper_cache_LDFLAGS =	\
+	$(SETTINGS_PLUGIN_LDFLAGS)	\
+        $(AM_LDFLAGS)
+
+gnome_update_wallpaper_cache_LDADD =	\
+	$(SETTINGS_PLUGIN_LIBS)
+
 EXTRA_DIST = 				\
 	$(Dbusapi_DATA)			\
 	gnome-settings-manager.xml	\
Index: gnome-settings-daemon-2.29.90/gnome-settings-daemon/gnome-update-wallpaper-cache.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ gnome-settings-daemon-2.29.90/gnome-settings-daemon/gnome-update-wallpaper-cache.c	2010-02-16 13:12:25.320343456 +0100
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2010 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License version 3.0 for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by Didier Roche <didrocks@ubuntu.com>
+ */
+
+#include <glib.h>
+#include <gdk/gdk.h>
+#include <gconf/gconf-client.h>
+#define GNOME_DESKTOP_USE_UNSTABLE_API
+#include <libgnomeui/gnome-bg.h>
+
+static GOptionEntry entries[] =
+{
+  { NULL }
+};
+
+main (int argc, char *argv[])
+{
+    GOptionContext *context = NULL;
+    GError         *error = NULL;
+
+    GdkScreen *screen;
+    GdkRectangle rect;
+    GnomeBG *bg;
+    GConfClient *client;
+    GdkPixbuf *pixbuf;
+
+    gdk_init (&argc, &argv);
+
+    context = g_option_context_new ("- refresh wallpaper cache");
+    g_option_context_add_main_entries (context, entries, NULL);
+    if (!g_option_context_parse (context, &argc, &argv, &error)) {
+        g_printerr ("option parsing failed: %s\n", error->message);
+        g_option_context_free(context);
+        g_error_free (error);
+        return (1);
+    }
+    if (context)
+        g_option_context_free (context);
+
+    /* cache only the first monitor */
+    screen = gdk_screen_get_default ();
+    gdk_screen_get_monitor_geometry (screen, 0, &rect);
+
+    bg = gnome_bg_new ();
+    client = gconf_client_get_default ();
+    gnome_bg_load_from_preferences (bg, client);
+
+    pixbuf = gdk_pixbuf_new (GDK_COLORSPACE_RGB, FALSE, 8, rect.width, rect.height);
+    gnome_bg_draw (bg, pixbuf, screen, FALSE);
+
+    return (0);
+}
