diff --git a/plugins/power/gsd-power-manager.c b/plugins/power/gsd-power-manager.c
index 3f10c22..5265ce7 100644
--- a/plugins/power/gsd-power-manager.c
+++ b/plugins/power/gsd-power-manager.c
@@ -646,7 +646,8 @@ engine_recalculate_state_icon (GsdPowerManager *manager)
         if (manager->priv->previous_icon == NULL) {
 
                 /* set fallback icon */
-                gtk_status_icon_set_visible (manager->priv->status_icon, TRUE);
+                if (g_strcmp0 (g_getenv ("XDG_CURRENT_DESKTOP"), "Unity") != 0)
+                        gtk_status_icon_set_visible (manager->priv->status_icon, TRUE);
                 gtk_status_icon_set_from_gicon (manager->priv->status_icon, icon);
                 manager->priv->previous_icon = icon;
                 return TRUE;
@@ -3338,6 +3339,8 @@ gsd_power_manager_start (GsdPowerManager *manager,
         /* TRANSLATORS: this is the title of the power manager status icon
          * that is only shown in fallback mode */
         gtk_status_icon_set_title (manager->priv->status_icon, _("Power Manager"));
+        if (g_strcmp0 (g_getenv ("XDG_CURRENT_DESKTOP"), "Unity") == 0)
+                gtk_status_icon_set_visible (manager->priv->status_icon, FALSE);
 
         /* connect to UPower for async power operations */
         g_dbus_proxy_new_for_bus (G_BUS_TYPE_SYSTEM,
