diff --git a/plugins/power/gsd-power-manager.c b/plugins/power/gsd-power-manager.c
index 3f10c22..2b4679d 100644
--- a/plugins/power/gsd-power-manager.c
+++ b/plugins/power/gsd-power-manager.c
@@ -646,7 +646,8 @@ engine_recalculate_state_icon (GsdPowerManager *manager)
         if (manager->priv->previous_icon == NULL) {
 
                 /* set fallback icon */
-                gtk_status_icon_set_visible (manager->priv->status_icon, TRUE);
+                if (g_strcmp0 (g_getenv ("XDG_CURRENT_DESKTOP"), "Unity") != 0)
+                        gtk_status_icon_set_visible (manager->priv->status_icon, TRUE);
                 gtk_status_icon_set_from_gicon (manager->priv->status_icon, icon);
                 manager->priv->previous_icon = icon;
                 return TRUE;
