Index: gnome-settings-daemon-3.3.5/plugins/mouse/gsd-mouse-manager.c
===================================================================
--- gnome-settings-daemon-3.3.5.orig/plugins/mouse/gsd-mouse-manager.c	2012-02-13 17:31:02.484427556 +0100
+++ gnome-settings-daemon-3.3.5/plugins/mouse/gsd-mouse-manager.c	2012-02-13 17:31:02.556427556 +0100
@@ -512,7 +512,7 @@
 
                 args[0] = "syndaemon";
                 args[1] = "-i";
-                args[2] = "2.0";
+                args[2] = "0.5";
                 args[3] = "-K";
                 args[4] = "-R";
                 args[5] = NULL;
