Index: gnome-settings-daemon-3.5.2/plugins/mouse/gsd-mouse-manager.c
===================================================================
--- gnome-settings-daemon-3.5.2.orig/plugins/mouse/gsd-mouse-manager.c	2012-06-05 17:25:28.200524607 +1200
+++ gnome-settings-daemon-3.5.2/plugins/mouse/gsd-mouse-manager.c	2012-06-05 17:25:31.228524502 +1200
@@ -577,7 +577,8 @@
                 args[2] = "1.0";
                 args[3] = "-K";
                 args[4] = "-R";
-                args[5] = NULL;
+                args[5] = "-t";                
+                args[6] = NULL;
 
                 if (!have_program_in_path (args[0]))
                         return 0;
