Index: gnome-settings-daemon-3.4.2/plugins/mouse/gsd-mouse-manager.c
===================================================================
--- gnome-settings-daemon-3.4.2.orig/plugins/mouse/gsd-mouse-manager.c	2012-06-18 10:26:32.759959310 +0100
+++ gnome-settings-daemon-3.4.2/plugins/mouse/gsd-mouse-manager.c	2012-06-18 10:27:04.615958743 +0100
@@ -542,7 +542,7 @@
 {
         if (state && touchpad_is_present ()) {
                 GError *error = NULL;
-                char *args[6];
+                char *args[7];
 
                 if (manager->priv->syndaemon_spawned)
                         return 0;
@@ -552,7 +552,8 @@
                 args[2] = "2.0";
                 args[3] = "-K";
                 args[4] = "-R";
-                args[5] = NULL;
+                args[5] = "-t";                
+                args[6] = NULL;
 
                 if (!have_program_in_path (args[0]))
                         return 0;
