Index: gnome-settings-daemon-3.4.2/plugins/mouse/gsd-mouse-manager.c
===================================================================
--- gnome-settings-daemon-3.4.2.orig/plugins/mouse/gsd-mouse-manager.c	2012-05-22 20:13:30.949505284 +0200
+++ gnome-settings-daemon-3.4.2/plugins/mouse/gsd-mouse-manager.c	2012-05-22 20:13:37.881505041 +0200
@@ -552,7 +552,8 @@
                 args[2] = "2.0";
                 args[3] = "-K";
                 args[4] = "-R";
-                args[5] = NULL;
+                args[5] = "-t";                
+                args[6] = NULL;
 
                 if (!have_program_in_path (args[0]))
                         return 0;
