--- a/plugins/mouse/gsd-mouse-manager.c
+++ b/plugins/mouse/gsd-mouse-manager.c
@@ -567,7 +567,7 @@
 {
         if (state && touchpad_is_present ()) {
                 GError *error = NULL;
-                char *args[6];
+                char *args[7];
 
                 if (manager->priv->syndaemon_spawned)
                         return 0;
@@ -577,7 +577,8 @@
                 args[2] = "1.0";
                 args[3] = "-K";
                 args[4] = "-R";
-                args[5] = NULL;
+                args[5] = "-t";                
+                args[6] = NULL;
 
                 if (!have_program_in_path (args[0]))
                         return 0;
