Debian #563592
GNOME #606421

Index: gnome-settings-daemon-2.32.1/plugins/housekeeping/gsd-disk-space.c
===================================================================
--- gnome-settings-daemon-2.32.1.orig/plugins/housekeeping/gsd-disk-space.c	2010-11-16 00:10:51.000000000 +1100
+++ gnome-settings-daemon-2.32.1/plugins/housekeeping/gsd-disk-space.c	2010-12-21 10:23:28.372750001 +1100
@@ -308,10 +308,13 @@
          * expose this in a way that allows it to be used for this
          * purpose
          */
+
+        /* We also ignore network filesystems. */
                  
         const gchar *ignore_fs[] = {
                 "auto",
                 "autofs",
+                "autofs4",
                 "devfs",
                 "devpts",
                 "ecryptfs",
@@ -328,6 +331,20 @@
                 "nfsd",
                 "rpc_pipefs",
                 "zfs",
+
+                "adfs",
+                "afs",
+                "cifs",
+                "cxfs",
+                "gfs",
+                "gfs2",
+                "lustre",
+                "lustre_lite",
+                "ncpfs",
+                "nfs",
+                "nfs4",
+                "ocfs2",
+                "smbfs",
                 NULL
         };
         const gchar *ignore_devices[] = {
