Index: gnome-settings-daemon-2.31.91/data/gnome-settings-daemon.pc.in
===================================================================
--- gnome-settings-daemon-2.31.91.orig/data/gnome-settings-daemon.pc.in	2010-09-01 19:01:44.722616999 +1000
+++ gnome-settings-daemon-2.31.91/data/gnome-settings-daemon.pc.in	2010-09-01 19:01:48.902617000 +1000
@@ -4,6 +4,7 @@
 includedir=@includedir@
 libexecdir=@libexecdir@
 plugindir=${libdir}/gnome-settings-daemon-2.0
+binary=${libexecdir}/gnome-settings-daemon
 
 Name: gnome-settings-daemon
 Description: Utility library for accessing gnome-settings-daemon over DBUS
