diff -Nur -x '*.orig' -x '*~' gnome-settings-daemon-2.27.4/data/gnome-settings-daemon.pc.in gnome-settings-daemon-2.27.4.new/data/gnome-settings-daemon.pc.in
--- gnome-settings-daemon-2.27.4/data/gnome-settings-daemon.pc.in	2009-05-03 19:15:30.000000000 +0100
+++ gnome-settings-daemon-2.27.4.new/data/gnome-settings-daemon.pc.in	2009-07-20 00:05:17.088827307 +0100
@@ -2,6 +2,8 @@
 exec_prefix=@exec_prefix@
 libdir=@libdir@
 includedir=@includedir@
+libexecdir=@libexecdir@
+binary=${libexecdir}/gnome-settings-daemon
 
 Name: gnome-settings-daemon
 Description: Utility library for accessing gnome-settings-daemon over DBUS
