Index: gnome-settings-daemon-2.30.2/data/gnome-settings-daemon.pc.in
===================================================================
--- gnome-settings-daemon-2.30.2.orig/data/gnome-settings-daemon.pc.in	2010-06-22 10:49:22.000000000 +0200
+++ gnome-settings-daemon-2.30.2/data/gnome-settings-daemon.pc.in	2010-07-23 00:51:03.373449845 +0200
@@ -4,6 +4,7 @@
 includedir=@includedir@
 libexecdir=@libexecdir@
 plugindir=@plugindir@
+binary=${libexecdir}/gnome-settings-daemon
 
 Name: gnome-settings-daemon
 Description: Utility library for accessing gnome-settings-daemon over DBUS
