
Index: gnome-settings-daemon-2.28.1/configure.ac
===================================================================
--- gnome-settings-daemon-2.28.1.orig/configure.ac	2010-04-09 00:06:20.434216920 +0200
+++ gnome-settings-daemon-2.28.1/configure.ac	2010-04-09 00:06:46.350216758 +0200
@@ -220,7 +220,7 @@ AC_CHECK_X_LIB(Xxf86misc, XF86MiscQueryE
 AC_SUBST(XF86MISC_LIBS)
 AC_CHECK_X_HEADERS([X11/extensions/XKB.h])
 
-PKG_CHECK_MODULES(LIBGNOMEKBD, [libgnomekbd >= 2.21.4 libxklavier >= 4.0])
+PKG_CHECK_MODULES(LIBGNOMEKBD, [libgnomekbd >= 2.21.4 libxklavier >= 5.0])
 AC_SUBST(LIBGNOMEKBD_CFLAGS)
 AC_SUBST(LIBGNOMEKBD_LIBS)
 
Index: gnome-settings-daemon-2.28.1/plugins/keyboard/gsd-keyboard-xkb.c
===================================================================
--- gnome-settings-daemon-2.28.1.orig/plugins/keyboard/gsd-keyboard-xkb.c	2010-04-09 00:07:01.986713878 +0200
+++ gnome-settings-daemon-2.28.1/plugins/keyboard/gsd-keyboard-xkb.c	2010-04-09 00:07:42.686216451 +0200
@@ -526,7 +526,7 @@ gsd_keyboard_xkb_shutdown (void)
 	if (!inited_ok)
 		return;
 
-	xkl_engine_stop_listen (xkl_engine);
+	xkl_engine_stop_listen (xkl_engine, XKLL_MANAGE_LAYOUTS | XKLL_MANAGE_WINDOW_STATES);
 
 	gdk_window_remove_filter (NULL,
 				  (GdkFilterFunc)
