--- a/plugins/common/gsd-keygrab.c
+++ b/plugins/common/gsd-keygrab.c
@@ -146,9 +146,9 @@ grab_key_unsafe (Key                 *ke
          * (which are useful to grab without a modifier).
          */
         if ((modifiers & gsd_used_mods) == 0 &&
-            (IN_RANGE(key->keysym, XF86KEYS_RANGE_MIN, XF86KEYS_RANGE_MAX) ||
-             IN_RANGE(key->keysym, FKEYS_RANGE_MIN, FKEYS_RANGE_MAX) ||
-             key->keysym == GDK_KEY_Pause)) {
+            !IN_RANGE(key->keysym, XF86KEYS_RANGE_MIN, XF86KEYS_RANGE_MAX) &&
+            !IN_RANGE(key->keysym, FKEYS_RANGE_MIN, FKEYS_RANGE_MAX) &&
+             key->keysym != GDK_KEY_Pause) {
                 GString *keycodes;
 
                 keycodes = g_string_new ("");
