Description: Show in Unity too
Author: Michael Terry <michael.terry@canonical.com>
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=654919
Bug-Ubuntu: https://launchpad.net/bugs/803519

Index: gnome-settings-daemon-3.1.3/data/gnome-settings-daemon.desktop.in.in
===================================================================
--- gnome-settings-daemon-3.1.3.orig/data/gnome-settings-daemon.desktop.in.in	2011-07-19 14:52:52.980082428 -0400
+++ gnome-settings-daemon-3.1.3/data/gnome-settings-daemon.desktop.in.in	2011-07-19 14:53:00.752082510 -0400
@@ -2,7 +2,7 @@
 Type=Application
 _Name=GNOME Settings Daemon
 Exec=@LIBEXECDIR@/gnome-settings-daemon
-OnlyShowIn=GNOME;
+OnlyShowIn=GNOME;Unity;
 NoDisplay=true
 X-GNOME-Autostart-Phase=Initialization
 X-GNOME-Autostart-Notify=true
