Index: gnome-settings-daemon-3.5.2/plugins/power/gsd-power-manager.c
===================================================================
--- gnome-settings-daemon-3.5.2.orig/plugins/power/gsd-power-manager.c	2012-06-05 17:32:11.184510655 +1200
+++ gnome-settings-daemon-3.5.2/plugins/power/gsd-power-manager.c	2012-06-05 17:32:13.628510569 +1200
@@ -688,7 +688,7 @@
 
         /* show a different icon if we are disconnected */
         icon = engine_get_icon (manager);
-        gtk_status_icon_set_visible (manager->priv->status_icon, icon != NULL);
+        gtk_status_icon_set_visible (manager->priv->status_icon, FALSE);
 
         if (icon == NULL) {
                 /* none before, now none */
@@ -704,6 +704,7 @@
         if (manager->priv->previous_icon == NULL) {
 
                 /* set fallback icon */
+                gtk_status_icon_set_visible (manager->priv->status_icon, FALSE);
                 gtk_status_icon_set_from_gicon (manager->priv->status_icon, icon);
                 manager->priv->previous_icon = icon;
                 return TRUE;
@@ -3704,6 +3705,7 @@
         /* TRANSLATORS: this is the title of the power manager status icon
          * that is only shown in fallback mode */
         gtk_status_icon_set_title (manager->priv->status_icon, _("Power Manager"));
+        gtk_status_icon_set_visible (manager->priv->status_icon, FALSE);
 
         /* connect to UPower for async power operations */
         g_dbus_proxy_new_for_bus (G_BUS_TYPE_SYSTEM,
