Description: Show in Unity too
Author: Michael Terry <michael.terry@canonical.com>
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=654919
Bug-Ubuntu: https://launchpad.net/bugs/803519

Index: gnome-settings-daemon-3.3.5/plugins/automount/gnome-fallback-mount-helper.desktop.in.in
===================================================================
--- gnome-settings-daemon-3.3.5.orig/plugins/automount/gnome-fallback-mount-helper.desktop.in.in	2011-10-17 19:36:56.000000000 +0200
+++ gnome-settings-daemon-3.3.5/plugins/automount/gnome-fallback-mount-helper.desktop.in.in	2012-02-13 17:31:02.632427556 +0100
@@ -7,6 +7,6 @@
 Type=Application
 Categories=
 NoDisplay=true
-OnlyShowIn=GNOME;
+OnlyShowIn=GNOME;Unity;
 X-GNOME-Autostart-Notify=true
-AutostartCondition=GNOME3 if-session gnome-fallback
+AutostartCondition=GNOME3 unless-session gnome
