Description: Show in Unity too
Author: Michael Terry <michael.terry@canonical.com>
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=654919
Bug-Ubuntu: https://launchpad.net/bugs/803519

--- a/plugins/automount/gnome-fallback-mount-helper.desktop.in.in
+++ b/plugins/automount/gnome-fallback-mount-helper.desktop.in.in
@@ -9,4 +9,4 @@
 NoDisplay=true
 OnlyShowIn=GNOME;Unity;
 X-GNOME-Autostart-Notify=true
-AutostartCondition=GNOME3 if-session gnome-fallback
+AutostartCondition=GNOME3 unless-session gnome
