Index: gnome-settings-daemon-3.3.91/plugins/xsettings/gsd-xsettings-manager.c
===================================================================
--- gnome-settings-daemon-3.3.91.orig/plugins/xsettings/gsd-xsettings-manager.c	2012-03-05 21:04:12.737981057 +0100
+++ gnome-settings-daemon-3.3.91/plugins/xsettings/gsd-xsettings-manager.c	2012-03-05 21:04:12.989981069 +0100
@@ -681,6 +681,7 @@
 
         for (i = 0; manager->priv->managers [i]; i++) {
                 xsettings_manager_set_int (manager->priv->managers [i], "Gtk/ShellShowsAppMenu", have_shell);
+                xsettings_manager_set_int (manager->priv->managers [i], "Gtk/ShellShowsMenubar", have_shell);
                 xsettings_manager_notify (manager->priv->managers [i]);
         }
         gnome_settings_profile_end (NULL);
@@ -891,7 +892,7 @@
 
         /* Shell flag */
         manager->priv->shell_name_watch_id = g_bus_watch_name (G_BUS_TYPE_SESSION,
-                                                               "org.gnome.Shell",
+                                                               "com.canonical.AppMenu.Registrar",
                                                                0,
                                                                on_shell_appeared,
                                                                on_shell_disappeared,
