Index: gnome-settings-daemon-3.3.5/plugins/xsettings/gsd-xsettings-manager.c
===================================================================
--- gnome-settings-daemon-3.3.5.orig/plugins/xsettings/gsd-xsettings-manager.c	2012-02-13 17:53:40.772436952 +0100
+++ gnome-settings-daemon-3.3.5/plugins/xsettings/gsd-xsettings-manager.c	2012-02-13 17:53:40.948436953 +0100
@@ -682,6 +682,7 @@
 
         for (i = 0; manager->priv->managers [i]; i++) {
                 xsettings_manager_set_int (manager->priv->managers [i], "Gtk/ShellShowsAppMenu", have_shell);
+                xsettings_manager_set_int (manager->priv->managers [i], "Gtk/ShellShowsMenubar", have_shell);
                 xsettings_manager_notify (manager->priv->managers [i]);
         }
         gnome_settings_profile_end (NULL);
@@ -892,7 +893,7 @@
 
         /* Shell flag */
         manager->priv->shell_name_watch_id = g_bus_watch_name (G_BUS_TYPE_SESSION,
-                                                               "org.gnome.Shell",
+                                                               "com.canonical.AppMenu.Registrar",
                                                                0,
                                                                on_shell_appeared,
                                                                on_shell_disappeared,
