From 2fe8419cc67644d3e83026dc4ad8094499b680e3 Mon Sep 17 00:00:00 2001
From: Jens Granseuer <jensgr@gmx.net>
Date: Fri, 30 Oct 2009 18:03:01 +0000
Subject: Honour libexecdir when spawning gsd-locate-pointer

Fixes bug #599209.
diff -Nur -x '*.orig' -x '*~' gnome-settings-daemon-2.28.1/plugins/mouse/gsd-mouse-manager.c gnome-settings-daemon-2.28.1.new/plugins/mouse/gsd-mouse-manager.c
--- gnome-settings-daemon-2.28.1/plugins/mouse/gsd-mouse-manager.c	2009-11-04 00:21:15.152736768 +0000
+++ gnome-settings-daemon-2.28.1.new/plugins/mouse/gsd-mouse-manager.c	2009-11-04 00:21:16.052735613 +0000
@@ -732,7 +732,7 @@
                 if (manager->priv->locate_pointer_spawned)
                         return;
 
-                args[0] = "/usr/libexec/gsd-locate-pointer";
+                args[0] = LIBEXECDIR "/gsd-locate-pointer";
                 args[1] = NULL;
 
                 g_spawn_async (NULL, args, NULL,
diff -Nur -x '*.orig' -x '*~' gnome-settings-daemon-2.28.1/plugins/mouse/Makefile.am gnome-settings-daemon-2.28.1.new/plugins/mouse/Makefile.am
--- gnome-settings-daemon-2.28.1/plugins/mouse/Makefile.am	2009-09-07 11:34:15.000000000 +0100
+++ gnome-settings-daemon-2.28.1.new/plugins/mouse/Makefile.am	2009-11-04 00:21:16.052735613 +0000
@@ -9,6 +9,7 @@
 libmouse_la_CPPFLAGS = \
 	-I$(top_srcdir)/gnome-settings-daemon		\
 	-DGNOME_SETTINGS_LOCALEDIR=\""$(datadir)/locale"\" \
+	-DLIBEXECDIR=\""$(libexecdir)"\" \
 	$(AM_CPPFLAGS)
 
 libmouse_la_CFLAGS = \
diff -Nur -x '*.orig' -x '*~' gnome-settings-daemon-2.28.1/plugins/mouse/Makefile.in gnome-settings-daemon-2.28.1.new/plugins/mouse/Makefile.in
--- gnome-settings-daemon-2.28.1/plugins/mouse/Makefile.in	2009-11-04 00:21:15.302735547 +0000
+++ gnome-settings-daemon-2.28.1.new/plugins/mouse/Makefile.in	2009-11-04 00:21:50.842736032 +0000
@@ -293,6 +293,7 @@
 libmouse_la_CPPFLAGS = \
 	-I$(top_srcdir)/gnome-settings-daemon		\
 	-DGNOME_SETTINGS_LOCALEDIR=\""$(datadir)/locale"\" \
+	-DLIBEXECDIR=\""$(libexecdir)"\" \
 	$(AM_CPPFLAGS)
 
 libmouse_la_CFLAGS = \
