Index: gnome-settings-daemon-3.4.0/plugins/power/gsd-power-manager.c
===================================================================
--- gnome-settings-daemon-3.4.0.orig/plugins/power/gsd-power-manager.c	2012-04-09 12:46:13.782227888 -0400
+++ gnome-settings-daemon-3.4.0/plugins/power/gsd-power-manager.c	2012-04-09 12:46:13.906227890 -0400
@@ -3863,6 +3863,8 @@
                 manager->priv->timeout_sleep_id = 0;
         }
 
+        g_signal_handlers_disconnect_by_data (manager->priv->up_client, manager);
+
         g_object_unref (manager->priv->session);
         g_object_unref (manager->priv->settings);
         g_object_unref (manager->priv->settings_screensaver);
Index: gnome-settings-daemon-3.4.0/plugins/updates/gsd-updates-refresh.c
===================================================================
--- gnome-settings-daemon-3.4.0.orig/plugins/updates/gsd-updates-refresh.c	2012-04-09 12:46:06.706227813 -0400
+++ gnome-settings-daemon-3.4.0/plugins/updates/gsd-updates-refresh.c	2012-04-09 12:46:13.910227890 -0400
@@ -616,6 +616,8 @@
         if (refresh->priv->force_get_updates_login_timeout_id != 0)
                 g_source_remove (refresh->priv->force_get_updates_login_timeout_id);
 
+        g_signal_handlers_disconnect_by_data (refresh->priv->client, refresh);
+
         g_object_unref (refresh->priv->control);
         g_object_unref (refresh->priv->settings);
         g_object_unref (refresh->priv->client);
Index: gnome-settings-daemon-3.4.0/plugins/xrandr/gsd-xrandr-manager.c
===================================================================
--- gnome-settings-daemon-3.4.0.orig/plugins/xrandr/gsd-xrandr-manager.c	2012-04-09 12:46:06.758227814 -0400
+++ gnome-settings-daemon-3.4.0/plugins/xrandr/gsd-xrandr-manager.c	2012-04-09 12:46:13.914227890 -0400
@@ -2026,6 +2026,7 @@
         }
 
         if (manager->priv->upower_client != NULL) {
+                g_signal_handlers_disconnect_by_data (manager->priv->upower_client, manager);
                 g_object_unref (manager->priv->upower_client);
                 manager->priv->upower_client = NULL;
         }
