--- a/plugins/keyboard/gsd-keyboard-manager.c
+++ b/plugins/keyboard/gsd-keyboard-manager.c
@@ -869,6 +869,7 @@
         const gchar *locale_variant = NULL;
         const gchar *locale;
         gchar *language;
+        GList *layouts;
 
         if (!layout)
                 return;
@@ -886,13 +887,17 @@
         if (!language)
                 language = language_code_from_locale (DEFAULT_LANGUAGE);
 
-        gnome_xkb_info_get_layout_info_for_language (manager->priv->xkb_info,
-                                                     language,
-                                                     NULL,
-                                                     NULL,
-                                                     NULL,
-                                                     &locale_layout,
-                                                     &locale_variant);
+        layouts = gnome_xkb_info_get_layouts_for_language (manager->priv->xkb_info,
+                                                           language);
+
+        if (layouts != NULL)
+          gnome_xkb_info_get_layout_info (manager->priv->xkb_info,
+                                          layouts->data,
+                                          NULL,
+                                          NULL,
+                                          &locale_layout,
+                                          &locale_variant);
+
         g_free (language);
 
         /* We want to minimize the number of XKB groups if we have
