--- a/plugins/keyboard/gsd-keyboard-manager.c
+++ b/plugins/keyboard/gsd-keyboard-manager.c
@@ -351,7 +351,7 @@
 
         g_variant_get (variant, "&s", &session_name);
 
-        if (g_strcmp0 (session_name, "gnome") == 0)
+        if (g_strcmp0 (session_name, "gnome") == 0 || g_strcmp0 (session_name, "ubuntu") == 0)
                 manager->priv->session_is_fallback = FALSE;
 
         g_variant_unref (variant);
