--- a/plugins/power/gsd-power-manager.c
+++ b/plugins/power/gsd-power-manager.c
@@ -4064,6 +4064,12 @@
                 return;
         }
 
+        /* Check session pointer as a proxy for whether the manager is in the
+           start or stop state */
+        if (manager->priv->session == NULL) {
+                return;
+        }
+
         g_debug ("Calling method '%s.%s' for Power",
                  interface_name, method_name);
 
@@ -4100,6 +4106,12 @@
 
         /* Check session pointer as a proxy for whether the manager is in the
            start or stop state */
+        if (manager->priv->session == NULL) {
+                return NULL;
+        }
+
+        /* Check session pointer as a proxy for whether the manager is in the
+           start or stop state */
         if (manager->priv->session == NULL) {
                 return NULL;
         }
