Index: gnome-settings-daemon-3.4.1/plugins/power/gsd-power-manager.c
===================================================================
--- gnome-settings-daemon-3.4.1.orig/plugins/power/gsd-power-manager.c	2012-04-25 16:53:47.385612908 -0400
+++ gnome-settings-daemon-3.4.1/plugins/power/gsd-power-manager.c	2012-04-25 17:04:19.765619601 -0400
@@ -4149,6 +4149,12 @@
 {
         GsdPowerManager *manager = GSD_POWER_MANAGER (user_data);
 
+        /* Check session pointer as a proxy for whether the manager is in the
+           start or stop state */
+        if (manager->priv->session == NULL) {
+                return;
+        }
+
         g_debug ("Calling method '%s.%s' for Power",
                  interface_name, method_name);
 
@@ -4183,6 +4189,12 @@
         GsdPowerManager *manager = GSD_POWER_MANAGER (user_data);
         GVariant *retval = NULL;
 
+        /* Check session pointer as a proxy for whether the manager is in the
+           start or stop state */
+        if (manager->priv->session == NULL) {
+                return NULL;
+        }
+
         if (g_strcmp0 (property_name, "Icon") == 0) {
                 retval = engine_get_icon_property_variant (manager);
         } else if (g_strcmp0 (property_name, "Tooltip") == 0) {
