Index: gnome-settings-daemon-3.4.2/configure.ac
===================================================================
--- gnome-settings-daemon-3.4.2.orig/configure.ac	2012-05-22 21:09:13.249387326 +0200
+++ gnome-settings-daemon-3.4.2/configure.ac	2012-05-22 21:09:23.933386948 +0200
@@ -57,10 +57,13 @@
 
 EXTRA_COMPILE_WARNINGS(yes)
 
+dnl FIXME we link against gconf-2.0 until all the plugins are ported
+dnl to GSettings, or we could get crashes on startup
 PKG_CHECK_MODULES(SETTINGS_DAEMON,
         glib-2.0 >= $GLIB_REQUIRED_VERSION
         gtk+-3.0 >= $GTK_REQUIRED_VERSION
         gio-2.0 >= $GIO_REQUIRED_VERSION
+        gconf-2.0 >= $GCONF_REQUIRED_VERSION
         gmodule-2.0
         gthread-2.0
 	gsettings-desktop-schemas >= 3.3.0
@@ -68,6 +71,7 @@
 
 PKG_CHECK_MODULES(SETTINGS_PLUGIN,
         gtk+-3.0 >= $GTK_REQUIRED_VERSION
+        gconf-2.0 >= $GCONF_REQUIRED_VERSION
         gio-2.0 >= $GIO_REQUIRED_VERSION
         gsettings-desktop-schemas
         x11
